design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,98 @@
+# Network Devices Capacity
+
+Author: [Edward Haas](edwardh@redhat.com)
+Contributors: TBD
+State: Draft
+
+## Summary
+Support the ability to define how many network devices a VM may potentially contain.
+Allowing the use of hotplug to add network interfaces up to that capacity limit.
+
+## Motivation
+The introduction of network devices hotplug introduced a limitation
+to the amount of possible devices one can attach.
+
+Users of libvirt who can estimate how many potential network devices are needed,
+can define the additional controllers in advance.
+KubeVirt does not expose such domain details through its API.
+But even if these details had been exposed by the VMI object,
+a common user would find it challenging to define such low domain details.
+
+### Goals
+- Allow users to define how many network devices can be potentially hotplug.
+- Do not expose domain architecture details (e.g. PCI/PCIe controllers).
+
+## Proposal
+Users are to express the potential amount of network devices through
+the VMI spec.
+Internally, kubevirt will use libvirt logic to generate the needed
+devices to support the amount specified (e.g. PCIe controllers).
+
+Using hotplug, users may add network devices up to the amount specified
+by the new user *interface-capacity* input.
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- TBD
+
+### API Extensions
+The VMI Spec will be extended to include a new field, named `interfaceCapacity`.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 1 and 32.
```

Comment:
can we really go up to 32 ?

Reply:
maybe also 0 is a valid value if soneone want to block adding ?

Reply:
I vaguely remember some 32 limit, but not really sure.
I will check with the libvirt devel.

> maybe also 0 is a valid value if soneone want to block adding ?

Blocking hotplug is not a goal or requirement.
I will explicitly mention it in a non-goals section.

By default libvirt provides the ability to hotplug an additional interface (or other PCIe device), and I am not interested in doing any changes in that regards. On the other hand, if someone does not specify this field at all, it is equivalent to specifying it as 0.
I'll change it to 0, to avoid interpreting "nothing" and 0 differently.

Reply:
> I vaguely remember some 32 limit, but not really sure. I will check with the libvirt devel.
> 
> > maybe also 0 is a valid value if soneone want to block adding ?
> 
> Blocking hotplug is not a goal or requirement. I will explicitly mention it in a non-goals section.
> 
> By default libvirt provides the ability to hotplug an additional interface (or other PCIe device), 

This is not exactly true; that only happens for q35 machine types. Meaning that depending on what is set on another part of the domain definition, you end up w/ a different behavior.

> and I am not interested in doing any changes in that regards. On the other hand, if someone does not specify this field at all, it is equivalent to specifying it as 0. I'll change it to 0, to avoid interpreting "nothing" and 0 differently.

Reply:
Why do you consider this a non-goal? It fits quite naturally into the design IMO.

Once this proposal is implemented, there are two possible scenarios:

* `networkInterfaceListCapacity` is not specified → existing behavior, the number of hotpluggable devices depends on the architecture and machine type and KubeVirt generally doesn't do anything to expose/control this value;
* `networkInterfaceListCapacity` is specified → KubeVirt can use the provided value to mediate all hotplug attempts.

In the latter scenario, it makes sense to validate coldplug scenarios as well. So for example

```yaml
spec:
  domain:
    devices:
      interfaces:
        - masquerade: {}
          name: default
        - bridge: {}
          name: secondary
      networkInterfaceListCapacity: 1
```

would be rejected outright because the user has asked for the number of interfaces to be 1 at most, but then provided a list of 2 elements, which clearly contradicts that.

Reply:
You make it sound very trivial, which it may be, but considering we also support the ability for this list to change dynamically with hot{un}plug, IMO we risk complicating things with no hard requirement.

We also defer validation logic like this one to the underlying implementation, i.e. libvirt. If an attempt is made beyond the available resources, it will fail.
We do not validate all details at the webhook level.

But while I think it is not a must, it can be added as an extension later.
Do you think it will be better to remove it from the non-goals and just add it as an optional extension?

Reply:
Yeah, I haven't actually looked at the code so I might be making incorrect assumptions about the complexity of implementing this :)

In my mental model, when a hot(un)plug request happens KubeVirt will need, even today, to at the very least perform the following steps:

* validate the provided information;
* generate an XML snippet for the new device;
* pass the XML to the appropriate libvirt API;
* update the corresponding list (`interfaces`, `disks`) to include the new device.

Assuming this mental model is roughly correct, implementing a check to ensure that the requested hot(un)plug operation will not cause the number of items in `interfaces` to grow larger than `networkInterfaceListCapacity` should be pretty straightforward.

Omitting the check and leaving validation to libvirt seems like a bad idea to me.

First of all, the resulting error message will probably look opaque and non-actionable to KubeVirt users. If the check happens in KubeVirt, we can provide much better feedback.

More importantly, and this is the concern that @AlonaKaplan has already expressed, the setting will not work intuitively: if you have 2 initial `interfaces` and set `networkInterfaceListCapacity: 3`, you expect a single hotplug operation to succeed, but libvirt will instead allow 2 hotplug operations for q35 and something like 30 for pc. I feel that this would result in the feature being perceived as ineffective/broken.

Can you please investigate whether performing this check really introduces a massive amount of additional complexity? I can also try looking into it myself to some extent, but I'm definitely going to need at least some pointers :)

Reply:
> Yeah, I haven't actually looked at the code so I might be making incorrect assumptions about the complexity of implementing this :)
> 
> In my mental model, when a hot(un)plug request happens KubeVirt will need, even today, to at the very least perform the following steps:
> 
> * validate the provided information;
> * generate an XML snippet for the new device;
> * pass the XML to the appropriate libvirt API;
> * update the corresponding list (`interfaces`, `disks`) to include the new device.
> 
> Assuming this mental model is roughly correct, implementing a check to ensure that the requested hot(un)plug operation will not cause the number of items in `interfaces` to grow larger than `networkInterfaceListCapacity` should be pretty straightforward.

Maybe there's something I'm missing, but, why does **un**plug matter here ? As I see it, this is only relevant for **add**ing interfaces, not when removing. 

> 
> Omitting the check and leaving validation to libvirt seems like a bad idea to me.
> 
> First of all, the resulting error message will probably look opaque and non-actionable to KubeVirt users. If the check happens in KubeVirt, we can provide much better feedback.
> 
> More importantly, and this is the concern that @AlonaKaplan has already expressed, the setting will not work intuitively: if you have 2 initial `interfaces` and set `networkInterfaceListCapacity: 3`, you expect a single hotplug operation to succeed, but libvirt will instead allow 2 hotplug operations for q35 and something like 30 for pc. I feel that this would result in the feature being perceived as ineffective/broken.

Exactly. It would be really hard IMO to argue the feature is working as expected under the circumstances defined above. I would personally prefer to define the number of controllers (my original proposal). That might be hard to understand / wrong abstraction but at least it didn't flat out lie to the user.

> 
> Can you please investigate whether performing this check really introduces a massive amount of additional complexity? I can also try looking into it myself to some extent, but I'm definitely going to need at least some pointers :)

Reply:
> Maybe there's something I'm missing, but, why does unplug matter here ? As I see it, this is only relevant for adding interfaces, not when removing.

No, you're right, validation only needs to happen when performing hotplug.

Reply:
> > Maybe there's something I'm missing, but, why does unplug matter here ? As I see it, this is only relevant for adding interfaces, not when removing.
> 
> No, you're right, validation only needs to happen when performing hotplug.

You are thinking about state machines, but this is all about desired state changes.
The spec changes, either to add or remove, and the spec is validated in its given state, without the knowledge of the previous state.

> In my mental model, when a hot(un)plug request happens KubeVirt will need, even today, to at the very least perform the following steps:
> 
> validate the provided information;

This step can occur at different locations, providing different feedbacks, having different costs.
The main validation point is at the webhook, which blocks changes immediately from being persisted into the API Server. We have already identify this location as problematic for containing logic and avoided (and even removed) checks.

> > Omitting the check and leaving validation to libvirt seems like a bad idea to me.
> > First of all, the resulting error message will probably look opaque and non-actionable to KubeVirt users. If the check happens in KubeVirt, we can provide much better feedback.

There is always a cost to be payed, it is indeed not free.
We can log this as an error close to the "attach" operation if it helps.
But I think there are other scenarios when this will fail and we will not know the reason, unless we add logic in kubevirt to examine the device configuration, something I want to avoid.

> > More importantly, and this is the concern that @AlonaKaplan has already expressed, the setting will not work intuitively: if you have 2 initial `interfaces` and set `networkInterfaceListCapacity: 3`, you expect a single hotplug operation to succeed, but libvirt will instead allow 2 hotplug operations for q35 and something like 30 for pc. I feel that this would result in the feature being perceived as ineffective/broken.
> 
> Exactly. It would be really hard IMO to argue the feature is working as expected under the circumstances defined above.

IMO you are adding a requirement which no one asked for.
No one asked a maximum capacity limit which needs to be enforced. What is required is the need to define a minimum capacity which needs to be enforced.
And I am interested to explicitly express this: Defining a capacity does not mean you are limited to it, you are free to plug until the backend rejects it. What the capacity does is to assure (as much as it can) that you are able to plug *at least* until that value.

I think kubevirt API should not be coupled with its backend libvirt, it should be free to evolve without this limitation.

> > Can you please investigate whether performing this check really introduces a massive amount of additional complexity? I can also try looking into it myself to some extent, but I'm definitely going to need at least some pointers :)

I still think we should defer this to a follow up addition and not require it as part of this proposal.
The discussion itself should be actually on the motivation and perhaps on the goals they trigger.
I see no motivation to define a maximum capacity nor to add logic in the API webhooks to enforce them.

I guess look at "capacity" in the same notion as the Go language does for list capacity. Or how how assured memory or assured bandwidth is defined.

It is even worth noting that this interface capacity is not really assured, it is used to express desired capacity, but in reality nothing can assure the backend devices are still there to serve it.

> I would personally prefer to define the number of controllers (my original proposal). That might be hard to understand / wrong abstraction but at least it didn't flat out lie to the user.

I suggest you comment on the alternative options and explain why it is better.
I fail to understand why defining a controller directly is better than letting the system/backend auto-calculate it. I just see over exposure of inner details and potential bugs from miss-configuration.

Reply:
> IMO you are adding a requirement which no one asked for.
> No one asked a maximum capacity limit which needs to be enforced. What is required is the need to define a minimum capacity which needs to be enforced.
> And I am interested to explicitly express this: Defining a capacity does not mean you are limited to it, you are free to plug until the backend rejects it. What the capacity does is to assure (as much as it can) that you are able to plug at least until that value.
> 
> I still think we should defer this to a follow up addition and not require it as part of this proposal.
> The discussion itself should be actually on the motivation and perhaps on the goals they trigger.
> I see no motivation to define a maximum capacity nor to add logic in the API webhooks to enforce them.
> 
> I guess look at "capacity" in the same notion as the Go language does for list capacity. Or how how assured memory or assured bandwidth is defined.
> 


Thanks for the explanation, I understand better where you're coming from now.

You're basically proposing something that has softer guarantees, and lower complexity, than what I had in mind.

I think that's perfectly fine, as long as

* the limitations are clearly explained in the documentation and, as much as possible, self-apparent from the API design;
* we don't paint ourselves in the corner when it comes to potentially introducing a stricter version in the future.

So I'm thinking maybe call the new value `interfacesListDesiredCapacity`, to highlight the fact that KubeVirt will do its best to allocate room for that many devices but there are no strict guarantees about it.


> I think kubevirt API should not be coupled with its backend libvirt, it should be free to evolve without this limitation.